<launch>

    <!--==========================-->
    <!-- Arguments and Parameters -->
    <!--==========================-->
    <arg name="debug" default="false"/>
    <arg name="use_clamp_arduino" default="false"/>
    <arg name="port_clamp" default="/dev/ttyACM0"/>
    <arg name="roll_radius" default="0.29"/>
    <arg name="clamp_scale_movement_up" default="0.5"/>
    <arg name="clamp_scale_movement_down" default="0.5"/>
    <arg name="clamp_scale_grasp" default="0.5"/>
    <arg name="target_x" default="0.0"/>
    <arg name="target_y" default="0.0"/>
    <arg name="target_frame" default="/odom"/>
    <arg name="target_tolerance" default="$(eval arg('roll_radius')*2)"/>
    <arg name="sensor_threshold" default="1.0"/>
    <arg name="grasp_angle" default="0.7854"/>
    <arg name="angle_tolerance" default="0.01"/>
    <arg name="K_angle" default="1.0"/>
    <arg name="K_linear" default="1.0"/>
    <arg name="max_velocity" default="0.25"/>
    <arg name="grasp_plate_offset_x" default="0.36"/>
    <arg name="grasp_plate_offset_y" default="-0.19"/>
    <arg name="approach_offset" default="2.0"/>
    <arg name="approach_tolerance" default="0.001"/>
    <arg name="backout_distance" default="1.0"/>
    <arg name="manual_deadman" default="4"/>
    <arg name="autonomous_deadman" default="5"/>
    <arg name="timeout" default="1.0"/>
    <arg name="sensor_frame" default="/lidar1_link"/>
    <arg name="point_cloud_topic" default="/os1_node1/points"/>


    <!--===============-->
    <!-- Clamp Control -->
    <!--===============-->
    <!-- The arduino running this code should be a MEGA and should be running "grasping/src/clamp_switch/clamp_switch.ino", see file for wiring -->
    <node if="$(arg use_clamp_arduino)" pkg="rosserial_python" name="clamp_node_arduino" type="serial_node.py" respawn="false">
        <param name="port" value="$(arg port_clamp)" />
        <param name="baud" value="19200"/>
    </node>

    <!-- Clamp Controller -->
    <node pkg="grasping" name="clamp_control" type="clamp_control" output="screen">
        <param name="clamp_scale_movement_up" value="$(arg clamp_scale_movement_up)"/>
        <param name="clamp_scale_movement_down" value="$(arg clamp_scale_movement_down)"/>
        <param name="clamp_scale_grasp" value="$(arg clamp_scale_grasp)"/>
        <remap from="/clamp_control/clamp_movement" to="/clamp_switch_node/clamp_movement"/>
        <remap from="/clamp_control/clamp_grasp" to="/clamp_switch_node/clamp_grasp"/>
    </node>

    <!--==================-->
    <!-- Approach Control -->
    <!--==================-->
    <node pkg="grasping" name="approach_controller" type="approach_controller" output="screen">
        <param name="target_x" value="$(arg target_x)"/>
        <param name="target_y" value="$(arg target_y)"/>
        <param name="/roll/radius" value="$(arg roll_radius)"/>
        <param name="grasp_angle" value="$(arg grasp_angle)"/>
        <param name="angle_tolerance" value="$(arg angle_tolerance)"/>
        <param name="K_angle" value="$(arg K_angle)"/>
        <param name="K_linear" value="1$(arg K_linear)"/>
        <param name="max_velocity" value="$(arg max_velocity)"/>
        <param name="grasp_plate_offset_x" value="$(arg grasp_plate_offset_x)"/>
        <param name="grasp_plate_offset_y" value="$(arg grasp_plate_offset_y)"/>
        <param name="approach_offset" value="$(arg approach_offset)"/>
        <param name="approach_tolerance" value="$(arg approach_tolerance)"/>
        <param name="backout_distance" value="$(arg backout_distance)"/>
        <param name="manual_deadman" value="$(arg manual_deadman)"/>
        <param name="autonomous_deadman" value="$(arg autonomous_deadman)"/>
        <param name="timeout" value="$(arg timeout)"/>
        <remap from="/approach_controller/point" to="/cylinder_detection/point"/>
    </node>

    <!--====================-->
    <!-- Cylinder Detection -->
    <!--====================-->
    <node pkg="grasping" name="cylinder_detection" type="cylinder_detection" respawn="true">
        <param name="debug" value="$(arg debug)"/>
        <param name="sensor_frame" value="$(arg sensor_frame)"/>
        <param name="point_cloud_topic" value="$(arg point_cloud_topic)"/>
        <param name="target_frame" value="$(arg target_frame)"/>
        <param name="target_x" value="$(arg target_x)"/>
        <param name="target_y" value="$(arg target_y)"/>
        <param name="target_tolerance" value="$(arg target_tolerance)"/>
        <param name="sensor_threshold" value="$(arg sensor_threshold)"/>
    </node>

</launch>
